### PR TITLE
Chart generator segment memory and new track combinations

### PR DESCRIPTION
**Pull request by dliappis**
Support new format of track combinations file and include new segment memory visualizations.

### CODE REVIEW

File: esrally/chart_generator.py [hunk 1]
```diff
@@ -51,11 +53,18 @@ def index_label(race_config):
     return label
 
 
-def format_title(environment, track_name, suffix=None):
+def format_title(chart_type, environment, track_name, suffix=None, flavor=None):
```
Comment (danielmitterdorfer) [inline esrally/chart_generator.py]: Nit: I think the suffix should always be the last parameter.

File: esrally/chart_generator.py [hunk 2]
```diff
@@ -51,11 +53,18 @@ def index_label(race_config):
...
+    title = "{}-{}".format(environment, track_name)
+
+    # as we generate two sets of Time Series/Nightly charts per flavor, the titles needs to be different.
+    if chart_type == TimeSeriesCharts:
```
Comment (danielmitterdorfer) [inline esrally/chart_generator.py]: This smells as if it is now in the wrong place if we need to do a type test. IMHO we should implement this now as a method on the respective chart classes.

File: esrally/chart_generator.py [hunk 3]
```diff
@@ -755,6 +769,9 @@ def index(environment, race_configs, title):
             }
         }
 
+    def __repr__(self):
```
Comment (danielmitterdorfer) [inline esrally/chart_generator.py]: Why is this needed?

File: esrally/chart_generator.py [hunk 4]
```diff
@@ -1187,6 +1319,9 @@ def index(environment, race_configs, title):
...
```
Comment (danielmitterdorfer) [inline esrally/chart_generator.py]: Why is this needed?

File: esrally/chart_generator.py [hunk 5]
```diff
@@ -1395,28 +1596,66 @@ def load_race_configs(cfg):
...
+        structures.append(dashboard)
+
+    return structures
+
+
+def gen_charts_from_track_combinations(race_configs, chart_type, env, logger):
+    structures = []
+    for flavor, race_configs_per_flavor in race_configs.items():
+        for race_configs_per_track in race_configs_per_flavor:
+            logger.debug("Generating charts for race_configs with name:[{}]/flavor:[{}]".format(race_configs_per_track[0].name, flavor))
```
Comment (danielmitterdorfer) [inline esrally/chart_generator.py]: Our convention is to use plain logger strings for log statements instead of the format API.

File: esrally/chart_generator.py [hunk 6]
```diff
@@ -1195,21 +1330,40 @@ def load_track(cfg, name=None):
     return track.load_track(cfg)
 
 
-def generate_index_ops(chart_type, race_configs, environment):
+def generate_index_ops(chart_type, race_configs, environment, logger):
     idx_race_configs = list(filter(lambda c: "indexing" in c.charts, race_configs))
+    for race_conf in idx_race_configs:
+        logger.debug("Gen index visualization for race config with name:[{}] / label:[{}] / flavor: {} / license: {}".format(
```
Comment (danielmitterdorfer) [inline esrally/chart_generator.py]: Our convention is to use plain logger strings for log statements instead of the format API.

File: esrally/chart_generator.py [hunk 7]
```diff
@@ -1195,21 +1330,40 @@ def load_track(cfg, name=None):
...
-        return [chart_type.index(environment, idx_race_configs, title)]
-    else:
-        return []
+        title = format_title(chart_type, environment, race_configs[0].track, "indexing-throughput", flavor=race_configs[0].flavor)
+        charts = [chart_type.index(environment, idx_race_configs, title)]
+    return charts
 
 
-def filter_string(environment, race_config):
+def filter_string_time_series(environment, race_config):
```
Comment (danielmitterdorfer) [inline esrally/chart_generator.py]: This should now be a method on the respective classes if we need to do type tests.

File: esrally/chart_generator.py [hunk 8]
```diff
@@ -1320,6 +1488,14 @@ def __init__(self, track, cfg=None, challenge=None, car=None, node_count=None, c
     def name(self):
         return self.configuration.get("name")
 
+    @property
+    def flavor(self):
+        return self.configuration.get("flavor")
+
+    @property
+    def lic(self):
```
Comment (danielmitterdorfer) [inline esrally/chart_generator.py]: What is `lic`? License? Is there a reason for the shortcut?

Comment (dliappis) [reply esrally/chart_generator.py]: Yes, it's license, but `license` is a built-in in Python: https://docs.python.org/3/library/builtins.html

Comment (danielmitterdorfer) [reply esrally/chart_generator.py]: 🤦‍♂️ - fair enough but does a property name really clash with a builtin? I mean, it's always used in the context of its enclosing object.

Comment (dliappis) [reply esrally/chart_generator.py]: It produces noise on the IDE at well and that was an additional motivation behind it. For variables at least it seems to be a [strong antipattern](https://stackoverflow.com/questions/16523789/naming-conflict-with-built-in-function) to shadow a built-in and @property in Python makes it act as an instance variable so didn't want to do this.

How about `eslicense`? Naming's not my strongest suit and yet it's hitting me all the time :)

Comment (danielmitterdorfer) [reply esrally/chart_generator.py]: Maybe `es_license` then?

File: esrally/chart_generator.py (hunk 2 above)
Comment (dliappis) [reply esrally/chart_generator.py]: The problem originally stemed from calling the static method `index` from within the floating function `generate_index_ops`:

https://github.com/elastic/rally/blob/c331b00faae9f544c7323ee2adeff5ea2f6390d8/esrally/chart_generator.py#L1341

I've refactored things in 5a3bedabf588681fce7aa0bb8d18789396447df9
